### ISSUE TITLE
Rule proposal: Check for unused propTypes

### ISSUE THREAD

**Issue by EvNaverniouk** [yannickcr/eslint-plugin-react#226]
While refactoring code you may remove the use of some props but forget to remove them from the propTypes. It would nice if ESLint could alert you with:

```
`foo` propType is defined but not used.
```

Similar to the unused vars rule.

**Comment by atomanyih** [comment]
:+1:

**Comment by winkler1** [comment]
Yes, this would be super useful

**Comment by epaulet** [comment]
:+1:

**Comment by mattdell** [comment]
Agree this would be incredibly useful :+1:

**Comment by jamesdwilson** [comment]
+1

**Comment by Jhawins** [comment]
+1 this seems necessary since we check for props missing a proptype definition we should also check for proptypes never being used

**Comment by onedesert** [comment]
👍

**Comment by ljharb** [comment]
All +1's should now be in the form of reactions on the first post - I'll clean up the ones here, and any future ones. Thanks for everyone's interest!

**Comment by EvNaverniouk** [comment]
I'm going to take a stab at writing this rule. Will submit a PR.

**Comment by lo1tuma** [comment]
What should happen if the whole `props` object is passed to an external third party function? Or if you are using the spread operator to pass-thru all props to a child component? I guess in such cases the rule should not warn.

**Comment by silvenon** [comment]
Yeah, unfortunately I think this rule is prone to false-positives.

**Comment by jamesdwilson** [comment]
Maybe we could just have it have options to issue warning or not if props object is passed to a third party function or using a spread operator? Instead of throwing the baby with the bathwater.

**Comment by EvNaverniouk** [comment]
I've done a first pass at this here: https://github.com/yannickcr/eslint-plugin-react/pull/611 (Instructions on how to test and install are in the PR).

Still hunting down false positives, but early feedback would be highly appreciated.

**Comment by mattdell** [comment]
Any update on this? Is there something preventing this from being merged?

**Comment by CrazyAlvaro** [comment]
@EvNaverniouk Thank you for add this rule here, Evgueni. During our development here, we found that this rule checks propType used under `this.props` but not under `nextProps`. In some cases we only need to update specific props by using `nextProps.foo`. Would it be nice if we could check on both `this.props` and `nextProps`?
